### PR TITLE
Ticket #1848: Add copy email button to all email fields in Django Admin

### LINKED ISSUE #1848: Copy email address onto clipboard
### Issue description

Problem: analyst need to copy an email address cleanly so that they can perform searches to verify the contact information through public sources. Sometimes, it's difficult to cleanly highlight and copy this email address, and it was a certain amount of time to get it correct.

Solution:Analysts want to have a BUTTON to copy an email address cleanly onto the clipboard so that it may be used in searches to verify the contact information through public sources.

### Acceptance criteria

- [ ] a button exists by each email address, which will copy the email address cleanly into the clipboard

### Additional context

See diagrams

### Links to other issues

_No response_

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #1848

## Changes

- Adds the `email_clipboard_change_form.html` change form template to all models that have the "email" field 
- Modifies the template for DomainInformation and DomainRequest to add a clipboard in the inline

## Context for reviewers
**This ticket requires #1852 + #1836 to be merged first. It contains changes from both of those. Please wait to review this item until then**

This PR adds a copy button to each email field in Django Admin. This means that all models should be checked.

## Setup
As both an analyst and admin, for all models in Django Admin click an individual record. Then, look for an email field. If one exists, there should be a clipboard next to it that allows you to copy the data inside of it.

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [x] Checked that the design translated visually
- [x] Checked behavior
- [x] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [x] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [x] Chrome
  - [x] Microsoft Edge
  - [ ] FireFox
  - [x] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots
image
image

### CODE REVIEW

File: src/registrar/templates/admin/fieldset.html [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+{% comment %}
+This is copied from Djangos implementation of this template, with added "blocks"
+It is not inherently customizable on its own, so we can modify this instead.
+https://github.com/django/django/blob/main/django/contrib/admin/templates/admin/includes/fieldset.html
+{% endcomment %}
+<fieldset class="module aligned {{ fieldset.classes }}">
+    {% block fieldset_title %}
+        {% if fieldset.name %}<h2>{{ fieldset.name }}</h2>{% endif %}
+    {% endblock fieldset_title %}
```
[the remark this answers is not in the document]
Comment (abroddrick) [reply src/registrar/templates/admin/fieldset.html:11]: I would add that comment in line here

File: src/registrar/templatetags/url_helpers.py [hunk 2]
```diff
@@ -26,6 +26,14 @@ def endswith(text, ends):
     return False
 
 
+@register.filter("split")
+def split_string(value, key):
```
Comment (abroddrick) [inline src/registrar/templatetags/url_helpers.py:30]: **general question** why is this added?

File: src/registrar/assets/js/get-gov-admin.js [hunk 3]
```diff
@@ -137,6 +137,83 @@ function openInNewTab(el, removeAttribute = false){
...
+            emailButtons.forEach((button) => {
+                button.addEventListener("click", ()=>{
+                    copyInnerTextToClipboard(button);
+                })
+            });
+        }
+    }
+
+    handleClipboardButtons();
+    handleClipboardLinks();
```
Comment (CocoByte) [inline src/registrar/assets/js/get-gov-admin.js:212]: I love how you organized this IIFE - just sayin'

File: src/registrar/assets/sass/_theme/_admin.scss [hunk 4]
```diff
@@ -416,8 +423,40 @@ address.dja-address-contact-list {
...
         color: var(--body-quiet-color);
     }
+
+    address button.usa-button__clipboard-link {
+        font-size: 0.875rem !important;
+    }
+
+    td button.usa-button__clipboard-link {
+        font-size: 0.875rem !important;
+    }
```
Comment (CocoByte) [inline src/registrar/assets/sass/_theme/_admin.scss:433]: You should be able to consolidate these into one declaration via a comma....while I debated whether we wanted to do this (there is merit to having separate declarations in some situations), it would make sense to consolidate since these are both instances of the same element.

try:
`address button.usa-button__clipboard-link, td button.usa-button__clipboard-link {
        font-size: 0.875rem !important;
    }
`

File: src/registrar/assets/sass/_theme/_admin.scss [hunk 5]
```diff
@@ -416,8 +423,40 @@ address.dja-address-contact-list {
...
 }
 
 .errors span.select2-selection {
     border: 1px solid var(--error-fg) !important;
+}
+
+.admin-icon-group {
+    position: relative;
```
Comment (CocoByte) [inline src/registrar/assets/sass/_theme/_admin.scss:441]: Could you add a comment up here saying this is all alignment-magic to make the copy button appear in the right place? (I'm assuming that's what this is, but it's not super clear just by reading the code)

**Review by CocoByte — COMMENTED** [review]
Minor adjustments -- not really blocking.  This one def needs designers to review.  Let me know if you need a quick re-view and I'll do a fast turnover for you

**Review by CocoByte — APPROVED** [review]
LGTM (just needs design review)

**Comment by gabydisarli** [comment]
Hey @zandercymatics I'll be doing the design review here. I like how we have the copy icon on the Contacts page. Can we replicate that where the email links are present but in a uneditable presentation, like on requests? Can we also add the "Copy" label to the button and show a little alert that tells the analysts that they've successfully copied the email to their clipboard, [similar to this prototype](https://mobbin.com/flows/551a2aab-2c5d-48d8-9a96-c338d1b1868c?tab=prototype)?

**Review by gabydisarli — APPROVED** [review]
Great work!
